README.md
# NanoCLUST

**De novo clustering and consensus building for ONT 16S sequencing data**.

## Introduction

The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a very portable manner. It comes with docker containers making installation trivial and results highly reproducible.

## Quick Start

i. Install [`nextflow`](https://nf-co.re/usage/installation)

ii. Install [`docker`](https://docs.docker.com/engine/installation/) or [`conda`](https://conda.io/miniconda.html)

iii. Clone the NanoCLUST repository and test the pipeline on a minimal dataset with a single command

*Download a BLAST database for cluster sequence classification. For NCBI 16S rRNA database:

```bash
mkdir db db/taxdb
wget https://ftp.ncbi.nlm.nih.gov/blast/db/16S_ribosomal_RNA.tar.gz && tar -xzvf 16S_ribosomal_RNA.tar.gz -C db
wget https://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz && tar -xzvf taxdb.tar.gz -C db/taxdb
```

```bash
nextflow run main.nf -profile test,<docker/conda>
```

iv. Start running your own analysis!

Run a single sample analysis using default parameters:

```bash
nextflow run nf-core/nanoclust -profile conda --reads 'sample.fastq' --db "path/to/nanoclust/db/16S_ribosomal_RNA" --tax "path/to/nanoclust/db/taxdb/"
```

See usage and output sections in the documentation (/docs) for all of the available options when running the pipeline.


## Credits

Rodríguez-Pérez H, Ciuffreda L, Flores C (2020). NanoCLUST: a species-level analysis of 16S rRNA nanopore sequencing data. Submitted.

This work was supported by Instituto de Salud Carlos III [PI14/00844, PI17/00610, and FI18/00230] and co-financed by the European Regional Development Funds, “A way of making Europe” from the European Union; Ministerio de Ciencia e Innovación [RTC-2017-6471-1, AEI/FEDER, UE]; Cabildo Insular de Tenerife [CGIEU0000219140]; Fundación Canaria Instituto de Investigación Sanitaria de Canarias [PIFUN48/18]; and by the agreement with Instituto Tecnológico y de Energías Renovables (ITER) to strengthen scientific and technological education, training, research, development and innovation in Genomics, Personalized Medicine and Biotechnology [OA17/008]. 

## Contributions and Support

If you would like to contribute to this pipeline, please see the contributing guidelines

For further information or help, don't hesitate to get in touch on [Slack](https://nfcore.slack.com/channels/nf-core/nanoclust) (you can join with [this invite](https://nf-co.re/join/slack)).


CHANGELOG.md
# nf-core/nanoclust: Changelog

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## v1.0dev - [date]

Initial release of nf-core/nanoclust, created with the [nf-core](http://nf-co.re/) template.

### `Added`

### `Fixed`

### `Dependencies`

### `Deprecated`


templates/kmer_freq.py
#!/usr/bin/env python

import sys
from Bio import SeqIO
from Bio.SeqIO.QualityIO import FastqGeneralIterator
from Bio.SeqIO.FastaIO import SimpleFastaParser
from collections import Counter,OrderedDict
from itertools import product,groupby
import math
import multiprocessing
import pandas as pd
from tqdm import tqdm
import argparse

def parse_args():
    # Create argument parser
    parser = argparse.ArgumentParser()

    # Positional mandatory arguments
    #parser.add_argument("fastx", help="Fasta/fastq file containing read sequences", type=str, default="$qced_reads")

    # Optional arguments
    parser.add_argument("-k", help="k-mer size [5]", type=int, default=5)
    parser.add_argument("-t", "--threads", help="Number of threads to use [4]", type=int, default=32)
    parser.add_argument("-c", "--count", help="Provide raw k-mer raw counts, not normalized [False]", action="store_true", default=False)
    parser.add_argument("-f", "--frac", help="Provide k-mer counts normalized by total number of k-mers [False]", action="store_true", default=False)

    # Parse arguments
    args = parser.parse_args()

    return args

def launch_pool( procs, funct, args ):
    p    = multiprocessing.Pool(processes=procs)
    try:
        results = p.map(funct, args)
        p.close()
        p.join()
    except KeyboardInterrupt:
        p.terminate()
    return results

def chunks( l, n ):
    """
    Yield successive n-sized chunks from l.
    """
    for i in range(0, len(l), n):
        yield l[i:i+n]

def rev_comp_motif( motif ):
    """
    Return the reverse complement of the input motif.
    """
    COMP = {"A":"T", \
            "T":"A", \
            "C":"G", \
            "G":"C", \
            "W":"S", \
            "S":"W", \
            "M":"K", \
            "K":"M", \
            "R":"Y", \
            "Y":"R", \
            "B":"V", \
            "D":"H", \
            "H":"D", \
            "V":"B", \
            "N":"N", \
            "X":"X", \
            "*":"*"}
    rc_motif = []
    for char in motif[::-1]:
        rc_motif.append( COMP[char] )
    return "".join(rc_motif)

def build_all_kmers( k ):
    kmers   = []
    for seq in product("ATGC",repeat=k):
        kmers.append( "".join(seq) )
    return kmers

def combine_kmers_list( all_kmers ):
    combined = set()
    for kmer in all_kmers:
        if rev_comp_motif(kmer) in combined:
            pass
        else:
            combined.add(kmer)
    combined = list(combined)
    combined.sort()
    return combined

def kmer_freq ( seq_str, k, combined_kmers, kmer_names_only=False ):
    seq_str    = seq_str.upper()
    
    all_kmer_n = Counter()
    for j in range( len(seq_str)-(k-1) ):
        motif = seq_str[j:j+k]
        all_kmer_n[motif] += 1

    # Combine forward and reverse complement motifs into one count
    combined_kmer_n = Counter()
    for kmer in combined_kmers:
        kmer_rc               = rev_comp_motif(kmer)
        combined_kmer_n[kmer] = all_kmer_n[kmer] + all_kmer_n[kmer_rc]
    return combined_kmer_n

def calc_seq_kmer_freqs( tup ):
    read_id        = tup[0]
    seq            = tup[1]
    k              = tup[2]
    combined_kmers = tup[3]
    i              = tup[4]
    count          = tup[5]
    frac           = tup[6]
    
    seq_comp            = []
    combined_kmer_n     = kmer_freq( seq, k, combined_kmers )
    ord_combined_kmer_n = OrderedDict(sorted(combined_kmer_n.items()))

    for kmer,n in ord_combined_kmer_n.items():
        if count:
            kmer_comp = n
        elif frac:
            kmer_comp = float(n) / sum(combined_kmer_n.values())
        else:
            kmer_comp = math.log(float(n + 1) / sum(combined_kmer_n.values())) # adding pseudocount for log transform
        seq_comp.append(kmer_comp)

    return read_id, seq_comp

def build_args_for_kmer_calc(read_num, target_range, args, read_id, seq, k, combined_kmers, lengths_d, count, frac):
    status = "keep going"
    if read_num>=target_range[0] and read_num<=target_range[1]:

        # if read_num%1000==0: print("Loading...",target_range, read_num)

        args.append( (read_id, seq, k, combined_kmers, read_num, count, frac) )
        lengths_d[read_id] = len(seq)
    elif read_num>target_range[1]:
        status = "over"
    return args,status

def launch_seq_kmers_pool( fastx, ftype, k, threads, target_range, combined_kmers, count, frac ):
    
    args      = []
    lengths_d = {}

    if ftype=="fastq":
        for read_num, (read_id, seq, qual) in enumerate(FastqGeneralIterator(open(fastx))):
            args,status = build_args_for_kmer_calc(read_num, target_range, args, read_id, seq, k, combined_kmers, lengths_d, count, frac)
            if status=="over":
                break

    elif ftype=="fasta":
        for read_num, (read_id, seq) in enumerate(SimpleFastaParser(open(fastx))):
            args,status = build_args_for_kmer_calc(read_num, target_range, args, read_id, seq, k, combined_kmers, lengths_d, count, frac)
            if status=="over":
                break
    
    results = launch_pool( threads, calc_seq_kmer_freqs, args )
    
    return dict(results), lengths_d

def print_comp_vectors(read_num, target_range, comp_vectors, read_id, lengths_d):
    status = "keep going"
    if read_num>=target_range[0] and read_num<=target_range[1]:

        # if read_num%1000==0: print("writing...",target_range, read_num)
        comp_vec_str = "\t".join( map(lambda x: str(round(x,4)), comp_vectors[read_id]))
        print("%s\t%i\t%s" % (read_id.split(" ")[0], lengths_d[read_id], comp_vec_str))
    elif read_num>target_range[1]:
        status = "over"
    return status

def write_output( fastx, ftype, comp_vectors, lengths_d, target_range ):

    if ftype=="fastq":
        for read_num, (read_id, seq, qual) in enumerate(FastqGeneralIterator(open(fastx))):
            status = print_comp_vectors(read_num, target_range, comp_vectors, read_id, lengths_d)
            if status=="over":
                break

    elif ftype=="fasta":
        for read_num, (read_id, seq) in enumerate(SimpleFastaParser(open(fastx))):
            status = print_comp_vectors(read_num, target_range, comp_vectors, read_id, lengths_d)
            if status=="over":
                break

def get_n_reads(fastx, ftype):
    n_lines = 0
    with open(fastx) as f:
        for i, l in enumerate(f):
            n_lines += 1
    
    if ftype=="fastq":
        n_reads = len([read_tup for read_tup in FastqGeneralIterator(open(fastx))])
    elif ftype=="fasta":
        n_reads = len([read_tup for read_tup in SimpleFastaParser(open(fastx))])
    return n_reads

def check_input_format(fastx):
    for line in open(fastx).readlines():
        break

    if line[0]=="@":
        ftype = "fastq"
    elif line[0]==">":
        ftype = "fasta"
    else:
        raise("Unexpected file type! Only *.fasta, *.fa, *.fsa, *.fna, *.fastq, and *.fq recognized.")
    return ftype

def main(args):
    ftype  = check_input_format("$qced_reads")
    n_reads = get_n_reads("$qced_reads", ftype)

    chunk_n_reads = 5000

    all_kmers = build_all_kmers(args.k)
    combined_kmers = combine_kmers_list(all_kmers)

    print("read\tlength\t%s" % "\t".join(combined_kmers))

    read_chunks = list(chunks(range(n_reads), chunk_n_reads))

    for chunk in tqdm(read_chunks):
        target_range = (chunk[0], chunk[-1])
        
        comp_vectors,lengths_d = launch_seq_kmers_pool( "$qced_reads",     \
                                                        ftype,          \
                                                        args.k,         \
                                                        args.threads,   \
                                                        target_range,   \
                                                        combined_kmers, \
                                                        args.count,     \
                                                        args.frac )
        write_output( "$qced_reads", ftype, comp_vectors, lengths_d, target_range )

if __name__=="__main__":
    args = parse_args()

    main(args)

templates/get_abundance.py
#!/usr/bin/env python

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rc
import pandas as pd
from functools import reduce
import requests
import json
#https://unipept.ugent.be/apidocs/taxonomy

def get_taxname(tax_id,tax_level):
    tags = {"S": "species_name","G": "genus_name","F": "family_name","O":'order_name', "C": "class_name"}
    tax_level_tag = tags[tax_level]
    
    path = 'http://api.unipept.ugent.be/api/v1/taxonomy.json?input[]=' + str(int(tax_id)) + '&extra=true&names=true'
    complete_tax = requests.get(path).text
    return json.loads(complete_tax)[0][tax_level_tag]

def get_abundance_values(names,paths):
    dfs = []
    for name,path in zip(names,paths):
        data = pd.read_csv(path, index_col=False, sep=';').iloc[:,1:]

        total = sum(data['reads_in_cluster'])
        rel_abundance=[]

        for index,row in data.iterrows():
            rel_abundance.append(row['reads_in_cluster'] / total)
            
        data['rel_abundance'] = rel_abundance
        dfs.append(pd.DataFrame({'taxid': data['taxid'], 'rel_abundance': rel_abundance}))
        data.to_csv("" + name + "_nanoclust_out.txt")

    return dfs

def merge_abundance(dfs,tax_level):
    df_final = reduce(lambda left,right: pd.merge(left,right,on='taxid',how='outer').fillna(0), dfs)
    df_final["taxid"] = [get_taxname(row["taxid"], tax_level) for index, row in df_final.iterrows()]
    df_final_grp = df_final.groupby(["taxid"], as_index=False).sum()
    return df_final_grp

def get_abundance(names,paths,tax_level):
    if(not isinstance(paths, list)):
        paths = [paths]
        names = [names]

    dfs = get_abundance_values(names,paths)
    df_final_grp = merge_abundance(dfs, tax_level)
    df_final_grp.to_csv("rel_abundance_"+ names[0] + "_" + tax_level + ".csv", index = False)

paths = "$table"
names = "$barcode"

get_abundance(names,paths, "G")
get_abundance(names,paths, "S")
get_abundance(names,paths, "O")
get_abundance(names,paths, "F")

templates/plot_abundances_pool.py
#!/usr/bin/env python

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rc
import pandas as pd

def plot_multi(table):

    df_final_grp = pd.read_csv(table)
    bars = [df_final_grp.iloc[i,1:].values.tolist() for i in range(0,df_final_grp.shape[0])]
    barWidth = 0.25
    colors = ['#a6cee3','#3caea3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c','#d11141','#ffc425','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928']
    r = np.arange(0, (len(df_final_grp.columns) -1))

    graph_bars = []
    graph_bars.append(plt.bar(r, bars[0], color=colors[0], edgecolor='white', width=barWidth, label=df_final_grp.iloc[0,0]))
    bottomm = bars[0]

    for index in range(1,len(bars)):
            graph_bars.append(plt.bar(r, bars[index], bottom=bottomm, color=colors[index], edgecolor='white', width=barWidth, label=df_final_grp.iloc[index,0]))
            bottomm = [i+j for i,j in zip(bottomm, bars[index])]

    plt.xticks(r, (df_final_grp.columns[1:]))
    plt.legend(graph_bars, df_final_grp["taxid"], ncol=3, fontsize='large',title_fontsize="x-large", loc='center',bbox_to_anchor=(0.5, -0.25))
    plt.savefig("" + table + ".plot.png", bbox_inches="tight")

table = "$table"

plot_multi(table)
plot_multi(table)
plot_multi(table)
plot_multi(table)

templates/.ipynb_checkpoints/umap_hdbscan-checkpoint.py
#!/usr/bin/env python

import numpy as np
import umap
import matplotlib.pyplot as plt
from sklearn import decomposition
import random
import pandas as pd
import hdbscan


df = pd.read_csv("$kmer_freqs", delimiter="\t")

#UMAP
motifs = [x for x in df.columns.values if x not in ["read", "length"]]
X = df.loc[:,motifs]
X_embedded = umap.UMAP(n_neighbors=15, min_dist=0.1, verbose=2).fit_transform(X)

df_umap = pd.DataFrame(X_embedded, columns=["D1", "D2"])
umap_out = pd.concat([df["read"], df["length"], df_umap], axis=1)

#HDBSCAN
X = umap_out.loc[:,["D1", "D2"]]
umap_out["bin_id"] = hdbscan.HDBSCAN(min_cluster_size=int($params.min_cluster_size), cluster_selection_epsilon=int($params.cluster_sel_epsilon)).fit_predict(X)
plt.figure(figsize=(30,19))
plt.scatter(X_embedded[:, 0], X_embedded[:, 1], c=umap_out["bin_id"], cmap='Spectral', s=0.7)
plt.gca().set_aspect('equal', 'datalim')
#plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
cluster_cnt = umap_out["bin_id"].max()
plt.colorbar(boundaries=np.arange(-1, cluster_cnt)-0.5).set_ticks(np.arange(-1, cluster_cnt))
plt.title('UMAP projection of the Digits dataset', fontsize=24);
plt.savefig('hdbscan.output.png')
umap_out.to_csv("output.hdbscan.tsv", sep="\t", index=False)


templates/umap_hdbscan.py
#!/usr/bin/env python

import numpy as np
import umap
import matplotlib.pyplot as plt
from sklearn import decomposition
import random
import pandas as pd
import hdbscan

df = pd.read_csv("$kmer_freqs", delimiter="\t")

#UMAP
motifs = [x for x in df.columns.values if x not in ["read", "length"]]
X = df.loc[:,motifs]
X_embedded = umap.UMAP(n_neighbors=15, min_dist=0.1, verbose=2).fit_transform(X)

df_umap = pd.DataFrame(X_embedded, columns=["D1", "D2"])
umap_out = pd.concat([df["read"], df["length"], df_umap], axis=1)

#HDBSCAN
X = umap_out.loc[:,["D1", "D2"]]
umap_out["bin_id"] = hdbscan.HDBSCAN(min_cluster_size=int($params.min_cluster_size), cluster_selection_epsilon=int($params.cluster_sel_epsilon)).fit_predict(X)

#PLOT
plt.figure(figsize=(20,20))
plt.scatter(X_embedded[:, 0], X_embedded[:, 1], c=umap_out["bin_id"], cmap='Spectral', s=1)
plt.xlabel("UMAP1", fontsize=18)
plt.ylabel("UMAP2", fontsize=18)
plt.gca().set_aspect('equal', 'datalim')
plt.title("Projecting " + str(len(umap_out['bin_id'])) + " reads. " + str(len(umap_out['bin_id'].unique())) + " clusters generated by HDBSCAN", fontsize=18)

for cluster in np.sort(umap_out['bin_id'].unique()):
    read = umap_out.loc[umap_out['bin_id'] == cluster].iloc[0]
    plt.annotate(str(cluster), (read['D1'], read['D2']), weight='bold', size=14)

plt.savefig('hdbscan.output.png')
umap_out.to_csv("hdbscan.output.tsv", sep="\t", index=False)
